### PR TITLE
fix: cache paths on Windows are broken

### PR DESCRIPTION
**Pull request by bartlomieju**
Cache paths are not properly generated on Windows.

It's broken since v0.13.0.

Ref: 
denoland/deno_std#552
denoland/deno_std#556

Example:
```rust
file:///C:/deno/js/unit_test_runner.ts
// for above URL this cache path should be generated 
file\c\deno\js\unit_test_runner.ts
// this is produced (probably)
C:/deno/js/unit_test_runner.ts
```

### CODE REVIEW

**Comment by bartlomieju** [comment]
@piscisaureus I think I managed to make it work. We're gonna need to do a release with this change to fix CI in standard lib

File: cli/disk_cache.rs [hunk 1]
```diff
@@ -37,13 +37,48 @@ impl DiskCache {
...
+          {
+            // Windows doesn't support ":" in filenames, so we need to extract disk prefix
+            // Example: file:///C:/deno/js/unit_test_runner.ts
+            // it should produce: file\c\deno\js\unit_test_runner.ts
+            match prefix_component.kind() {
+              Prefix::Disk(disk_byte) | Prefix::VerbatimDisk(disk_byte) => {
+                let disk = (disk_byte as char).to_string();
+                out.push(disk);
+              }
+              _ => {}
```
Comment (piscisaureus) [inline cli/disk_cache.rs]: Add an unreachable() here. Let's not silently discard part components.

Comment (bartlomieju) [reply cli/disk_cache.rs]: @piscisaureus added

**Comment by piscisaureus** [comment]
This fixes a very acute problem, so I'm all for landing this.
However, we have similar problems with other path components, which can also contain illegal characters (e.g. `?` and `&`).

Tangentially related, there's also a correctness issue w.r.t. case sensitivity. In URLs the path and query component are case sensitive, while the protocol and host are not. When deno_dir lives on a case insensitive file system it won't be able to handle that properly.

**Comment by bartlomieju** [comment]
> However, we have similar problems with other path components, which can also contain illegal characters (e.g. `?` and `&`).

Haven't thought about it, but I believe `Url` won't return them from `path_segments` method. That inherently means that `http://example.com/a.js?foo` would be cached as `http/example.com/a.js`.


> Tangentially related, there's also a correctness issue w.r.t. case sensitivity. In URLs the path and query component are case sensitive, while the protocol and host are not. When deno_dir lives on a case insensitive file system it won't be able to handle that properly.

Agreed, any ideas how to tackle that? Cast all protocols and hostnames to lowercase?

**Comment by piscisaureus** [comment]
> Haven't thought about it, but I believe `Url` won't return them from `path_segments` method. That inherently means that `http://example.com/a.js?foo` would be cached as `http/example.com/a.js`.

That doesn't seem right either. Only the part after `#` can be safely discarded.

**Comment by bartlomieju** [comment]
> That doesn't seem right either. Only the part after `#` can be safely discarded.

Agreed, shall I open a new issue with two described edge cases and we can land this patch?

**Pull request merged**